lib/versions/npm.js

```diff
@@ -1,82 +1,57 @@
-      }
-      else{
-        // otherwise, find our download complete measurement
-        entry = _.last(_.where(items.getEntries(), {name: downloadComplete}));
-        
-        if(entry){
-          strDuration = __('Finished downloading {{packageName}} {{version}} in {{duration}}ms', {packageName: packageName, version: version, duration: entry.duration});
-          performance.clearMarks();
-        }
-      }
-      
-      // log our measurement and make it red if it has taken too long
-      if(entry && strDuration){
-        if(entry.duration > 4000){
-          strDuration = strDuration.red;
-        }
-        this.logger.info(strDuration);
-      }
-      
-    });
-    obsMeasure.observe({entryTypes: ['measure']});
+  getPackageVersion(packageName, version, callback) {
+    const packagePath = Npm.getPackagePath(packageName, version);
 
-    // define mark and measurement names
-    let startMark = 'downloadStart' + packageName + version;
-    let ongoingMark = 'downloadOngoingMark' + packageName + version;
-    let downloadOngoing = 'downloadOngoing' + packageName + version;
-    let endMark = 'downloadEnd' + packageName + version;
-    let downloadComplete = 'downloadComplete' + packageName + version;
+    // check if this package already exists in the filesystem
+    if (fs.existsSync(packagePath)) {
+      return callback(null, packagePath);
+    }
 
-    // mark our start time
-    performance.mark(startMark);
+    const pluginManager = new PluginManager({pluginsPath: './.embark/versions/' + packageName + '/' + version + '/'});
     
-    // function that continually updates the console to show user that we're downloading a library
-    let intOngoingDownload = setInterval(
-      function(){ 
-        performance.mark(ongoingMark);
-        performance.measure(downloadOngoing, startMark, ongoingMark); 
-      }, 750);
+    // check if we're already installing this package
+    if(this._isInstalling(packageName, version)){
+      this._installing[packageName + version].push(callback);
+    }else{
+      this._installing[packageName + version] = new Array(callback);
 
-    // do the package download/install
-    manager.install(packageName, version).then((result) => {
-      // stop updating console for ongoing download
-      clearInterval(intOngoingDownload);
-      performance.mark(endMark);
-      performance.measure(downloadComplete, startMark, endMark);
-      callback(null , result.location);
-    }).catch(callback);
+      
+      const timer = new NpmTimer({logger: this._logger, packageName: packageName, version: version});
+      timer.start();
+
+      // do the package download/install
+      pluginManager.install(packageName, version).then((result) => {        
+          timer.end();
+          this._installing[packageName + version].forEach((cb) => {
+            cb(null, result.location);
+          });
+          delete this._installing[packageName + version];
+          //callback(null, result.location);
```

Comment:
✂️

Reply:
Updated to call all callbacks for that library in event of an error